### PR TITLE
[Inductor][CPP] Leverage full bits for BF16/FP16 vectorization

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133830
* __->__ #126502

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @peterbell10

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -3436,7 +3445,10 @@ def is_lowp_fp_scheduler(self, scheduler_node: SchedulerNode):
             return True
         # Propagate the dtype to check if all the fx node is bf16/fp16
         DataTypePropagation.propagate_scheduler_node(scheduler_node)
-        return get_loop_body_lowp_fp(scheduler_node._body) is not None
+        return (
+            get_loop_body_lowp_fp(scheduler_node._body)[0] is not None
+            and not get_loop_body_lowp_fp(scheduler_node._body)[1]
+        )
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:3451]: I guess we do not this extra check with the "full-bit" support in this PR? The original motivation of having this check is that we want to use "full-bit" as much as possible. Now we use "full-bit" by default. We can check whether to add type conversion inside the legalization logic not before it. We can do this cleanup in the next PR.

File: test/inductor/test_cpu_repro.py [hunk 2]
```diff
@@ -4214,6 +4214,90 @@ def fn(x1, x2, x3):
...
+    @requires_vectorization
+    def test_full_bits_lowp(self):
+        from torch._dynamo.utils import detect_fake_mode
+        from torch._inductor.compile_fx import shape_env_from_inputs
+        from torch._inductor.debug import DebugContext
+        from torch._inductor.graph import GraphLowering
+        from torch._inductor.virtualized import V
+        from torch.fx.passes.fake_tensor_prop import FakeTensorProp
+
+        def check_use_full_bits(func, shapes, dtype, mixed, check_vecn):
```
Comment (jgong5) [inline test/inductor/test_cpu_repro.py:4226]: Can we use `run_and_get_cpp_code` to simplify the code?

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (CaoE) [reply torch/_inductor/codegen/cpp.py:3451]: Yes, we can check whether to add type conversion inside the legalization logic. Will do this in the subsequent PRs.

File: test/inductor/test_cpu_repro.py (hunk 2 above)
Comment (CaoE) [reply test/inductor/test_cpu_repro.py:4219]: Use `run_and_get_cpp_code ` instead.

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**